File: src/main/java/com/github/theonlytails/ruby/items/CustomSpawnEggItem.java
package com.github.theonlytails.ruby.items;

import net.minecraft.block.DispenserBlock;
import net.minecraft.dispenser.DefaultDispenseItemBehavior;
import net.minecraft.dispenser.IBlockSource;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.SpawnReason;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.SpawnEggItem;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.util.Direction;
import net.minecraftforge.common.util.Lazy;
import net.minecraftforge.fml.RegistryObject;
import net.minecraftforge.fml.common.ObfuscationReflectionHelper;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class CustomSpawnEggItem extends SpawnEggItem {
    private static final List<CustomSpawnEggItem> UNADDED_EGGS = new ArrayList<>();
    private final Lazy<? extends EntityType<?>> entityTypeSupplier;

    public CustomSpawnEggItem(final RegistryObject<? extends EntityType<?>> entityTypeSupplier, int primaryColorIn, int secondaryColorIn, Item.Properties builder) {
        //noinspection ConstantConditions
        super(null, primaryColorIn, secondaryColorIn, builder);
        this.entityTypeSupplier = Lazy.of(entityTypeSupplier);
        UNADDED_EGGS.add(this);

    }

    public static void initSpawnEggs() {
        final @NotNull Map<EntityType<?>, CustomSpawnEggItem> EGGS = Objects.requireNonNull(
                ObfuscationReflectionHelper.getPrivateValue(SpawnEggItem.class, null, "field_195987_b"));

        DefaultDispenseItemBehavior dispenserBehavior = new DefaultDispenseItemBehavior() {
            @Override
            protected @NotNull ItemStack dispenseStack(@NotNull IBlockSource source, @NotNull ItemStack stack) {
                Direction direction = source.getBlockState().get(DispenserBlock.FACING);
                EntityType<?> type = ((SpawnEggItem) stack.getItem()).getType(stack.getTag());
                type.spawn(source.getWorld(), stack, null, source.getBlockPos().offset(direction),
                        SpawnReason.DISPENSER, direction != Direction.UP, false);
                stack.shrink(1);
                return stack;
            }
        };

        for (final SpawnEggItem spawnEgg : UNADDED_EGGS) {
            EGGS.put(spawnEgg.getType(null), (CustomSpawnEggItem) spawnEgg);
            DispenserBlock.registerDispenseBehavior(spawnEgg, dispenserBehavior);
        }

        UNADDED_EGGS.clear();
    }

    @Override
    public @NotNull EntityType<?> getType(@Nullable CompoundNBT nbt) {
        return this.entityTypeSupplier.get();
    }
}

File: src/main/kotlin/com/github/theonlytails/rubymod/items/CustomSpawnEggItem.kt
package com.github.theonlytails.rubymod.items

import net.minecraft.block.DispenserBlock
import net.minecraft.dispenser.DefaultDispenseItemBehavior
import net.minecraft.dispenser.IBlockSource
import net.minecraft.entity.EntityType
import net.minecraft.entity.SpawnReason
import net.minecraft.item.ItemStack
import net.minecraft.item.SpawnEggItem
import net.minecraft.nbt.CompoundNBT
import net.minecraft.util.Direction
import net.minecraftforge.common.util.Lazy
import net.minecraftforge.fml.common.ObfuscationReflectionHelper
import java.util.ArrayList
import java.util.Objects
import java.util.function.Supplier

class CustomSpawnEggItem(
	entityTypeSupplier: Supplier<out EntityType<*>>,
	primaryColorIn: Int,
	secondaryColorIn: Int,
	builder: Properties,
) :
	SpawnEggItem(null, primaryColorIn, secondaryColorIn, builder) {
	private val entityTypeSupplier: Lazy<out EntityType<*>> = Lazy.of(entityTypeSupplier::get)
	override fun getType(nbt: CompoundNBT?): EntityType<*> {
		return entityTypeSupplier.get()
	}

	companion object {
		private val UNADDED_EGGS: MutableList<CustomSpawnEggItem> = ArrayList()
		fun initSpawnEggs() {
			val eggs: MutableMap<EntityType<*>, CustomSpawnEggItem>? = Objects.requireNonNull(
				ObfuscationReflectionHelper.getPrivateValue<MutableMap<EntityType<*>, CustomSpawnEggItem>, SpawnEggItem>(
					SpawnEggItem::class.java, null, "field_195987_b"))

			val dispenserBehavior: DefaultDispenseItemBehavior = object : DefaultDispenseItemBehavior() {
				override fun dispenseStack(source: IBlockSource, stack: ItemStack): ItemStack {
					val direction = source.blockState.get(DispenserBlock.FACING)
					val type = (stack.item as SpawnEggItem).getType(stack.tag)
					type.spawn(source.world, stack, null, source.blockPos.offset(direction),
						SpawnReason.DISPENSER, direction != Direction.UP, false)
					stack.shrink(1)
					return stack
				}
			}
			for (spawnEgg in UNADDED_EGGS) {
				eggs?.set(spawnEgg.getType(null), spawnEgg)
				DispenserBlock.registerDispenseBehavior(spawnEgg, dispenserBehavior)
			}
			UNADDED_EGGS.clear()
		}
	}

	init {
		UNADDED_EGGS.add(this)
	}
}